visualizer/src/Canvas/Labeled/LabeledCanvas.js

```diff
@@ -1,104 +1,66 @@
 import { useSelector } from '@xstate/react';
-import React, { useEffect, useRef } from 'react';
+import { GPU } from 'gpu.js';
+import { useEffect, useRef } from 'react';
 import { useCanvas, useFeature, useLabeled, useSelect } from '../../ProjectContext';
-import { highlightImageData, opacityImageData } from '../canvasUtils';
 
-export const LabeledCanvas = ({ className }) => {
-  const canvas = useCanvas();
-  const sx = useSelector(canvas, (state) => state.context.sx);
-  const sy = useSelector(canvas, (state) => state.context.sy);
-  const zoom = useSelector(canvas, (state) => state.context.zoom);
-  const scale = useSelector(canvas, (state) => state.context.scale);
-  const sw = useSelector(canvas, (state) => state.context.width);
-  const sh = useSelector(canvas, (state) => state.context.height);
+const highlightColor = [255, 0, 0];
 
-  const width = sw * scale * window.devicePixelRatio;
-  const height = sh * scale * window.devicePixelRatio;
+export const LabeledCanvas = ({ setCanvases }) => {
+  const canvas = useCanvas();
+  const width = useSelector(canvas, (state) => state.context.width);
+  const height = useSelector(canvas, (state) => state.context.height);
 
   const labeled = useLabeled();
   const featureIndex = useSelector(labeled, (state) => state.context.feature);
   const highlight = useSelector(labeled, (state) => state.context.highlight);
   const opacity = useSelector(labeled, (state) => state.context.opacity);
 
   const feature = useFeature(featureIndex);
-  const labeledImage = useSelector(feature, (state) => state.context.labeledImage);
+  const colormap = useSelector(feature, (state) => state.context.colormap);
   let labeledArray = useSelector(feature, (state) => state.context.labeledArray);
   if (!labeledArray) {
-    labeledArray = Array(sh).fill(Array(sw).fill(0));
+    labeledArray = Array(height * width).fill(0);
   }
 
   const select = useSelect();
   const foreground = useSelector(select, (state) => state.context.foreground);
-  const background = useSelector(select, (state) => state.context.background);
 
+  const kernel = useRef();
   const canvasRef = useRef();
-  const ctx = useRef();
-  const hiddenCanvasRef = useRef();
-  const hiddenCtx = useRef();
 
   useEffect(() => {
-    ctx.current = canvasRef.current.getContext('2d');
-    ctx.current.imageSmoothingEnabled = false;
+    const canvas = canvasRef.current;
+    const gl = canvas.getContext('webgl2', { premultipliedAlpha: false });
```

Comment:
What does the webgl2 context mean?

Reply:
There are several context types available for a canvas, including '2d', 'webgl', and 'webgl2'([https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext))

This line specifies the premultipliedAlpha setting for the webgl2 context of the canvas. We need it for the LabeledCanvas, OutlineCanvas, BrushCanvas, and ThresholdCanvas as they can include transparent/translucent pixels. Without this setting, composing the canvases block the raw image underneath. It's after this example in the gpu.js documentation: [https://github.com/gpujs/gpu.js/#alpha](https://github.com/gpujs/gpu.js/#alpha)